pkg/aggregator/time_sampler.go

```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
 
 		// Add sample to bucket
 		if err := bucketMetrics.AddSample(contextKey, metricSample, timestamp, s.interval, nil); err != nil {
-			log.Debugf("Ignoring sample '%s' on host '%s' and tags '%s': %s", metricSample.Name, metricSample.Host, metricSample.Tags, err)
+			log.Debugf("TimeSampler #%d Ignoring sample '%s' on host '%s' and tags '%s': %s", s.id, metricSample.Name, metricSample.Host, metricSample.Tags, err)
 		}
 	}
 }
 
+// Stop stops the time sampler. It can't be re-used after being stop,
+// use NewTimeSampler instead.
+func (s *TimeSampler) Stop() {
+	s.stopChan <- struct{}{}
+}
+
+// We process all receivend samples in the `select`, but we also process a flush action,
+// meaning that the time sampler will not process any sample while it is flushing.
+// Note that it was the same design in the BufferedAggregator (but at the aggregator level,
+// not sampler level).
+// If we want to move to a design where we can flush while we are processing samples,
+// we could consider implementing double-buffering or locking for every sample reception.
+func (s *TimeSampler) processLoop(flushInterval time.Duration) {
+	ticker := time.NewTicker(flushInterval)
+	for {
+		select {
+		case <-s.stopChan:
+			return
+		case ms := <-s.samples:
+			// do this telemetry here and not in the samplers goroutines, this way
+			// they won't compete for the telemetry locks.
+			aggregatorDogstatsdMetricSample.Add(int64(len(ms)))
+			tlmProcessed.Add(float64(len(ms)), "dogstatsd_metrics")
+			t := timeNowNano()
+			for i := 0; i < len(ms); i++ {
+				s.sample(&ms[i], t)
+			}
+			s.metricSamplePool.PutBatch(ms)
+		case command := <-s.FlushChan:
+			s.triggerFlush(command.Time, command.BlockChan != nil)
+			if command.BlockChan != nil {
+				command.BlockChan <- struct{}{}
+			}
+		case t := <-ticker.C:
+			s.triggerFlush(t, false)
+		}
+	}
+}
+
+func (s *TimeSampler) triggerFlush(t time.Time, waitForSerializer bool) {
+	if s.parallelSerialization.enabled {
+		s.triggerFlushWithParallelSerialize(t, waitForSerializer)
+	} else {
+		log.Debugf("Time Sampler #%d Flushing series to the forwarder", s.id)
+		var series metrics.Series
+		sketches := s.flush(float64(t.Unix()), &series) // XXX(remy): is this conversation correct? note that it is in second
```

Comment:
This looks correct, since we truncate it later anyway, though personally I'd keep the original `UnixNano()/time.Second` here, or just convert everything to use int64. Floating point timestamps look like they should be subsecond precision.